pkg/tarutil/tarutil.go

```diff
@@ -164,17 +165,29 @@ func ExtractFiles(r io.Reader, filenameMatcher matcher.Matcher) (LayerFiles, err
 		case tar.TypeReg, tar.TypeLink:
 			var fileData FileData
 
+			executable, _ := executableMatcher.Match(filename, hdr.FileInfo(), contents)
 			if hdr.Size > maxELFExecutableFileSize {
 				log.Warnf("Skipping ELF executable check for file %q (%d bytes) because it is larger than the configured maxELFExecutableFileSizeMB of %d", filename, hdr.Size, maxELFExecutableFileSize/1024/1024)
-			} else if hdr.Size >= elfHeaderSize { // Only bother attempting to get ELF metadata if the file is large enough for the ELF header.
-				fileData.ELFMetadata, err = elf.GetExecutableMetadata(contents)
-				if err != nil {
-					log.Errorf("Failed to get dependencies for %s: %v", filename, err)
+			} else {
+				if hdr.Size >= elfHeaderSize { // Only bother attempting to get ELF metadata if the file is large enough for the ELF header.
+					fileData.ELFMetadata, err = elf.GetExecutableMetadata(contents)
+					if err != nil {
+						log.Errorf("Failed to get dependencies for %s: %v", filename, err)
+					}
+				}
+				if executable && fileData.ELFMetadata == nil {
+					shebangBytes := make([]byte, 2)
+					if hdr.Size > 2 {
```

Comment:
nit: instead of 2, do len(shebangHeader) or maybe make a global var that is set to `len(shebangHeader)`